File: sq/core/geom/2d/Path.ts
import { Texture2d } from "script/engine/sq/assets";
import { Vec2 } from "../../math";
import Capsule from "./Capsule";
import Circle from "./Circle";
import Ellipse from "./Ellipse";
import Polygon from "./Polygon";
import Rectangle from "./Rectangle";
import ShapePrimitive from "./ShapePrimitive";

export default class Path extends ShapePrimitive {
    private shapePrimitives: Array<ShapePrimitive> = [];
    private _currentPolygonShape: ShapePrimitive;
    private dirty: boolean;
    private addPoint(x: number, y: number) {
        this._currentPolygonShape.points.push(new Vec2(x, y));
    }

    public moveTo(x: number, y: number) {
        this.dirty = true;
        this._currentPolygonShape = new Polygon();
        this.shapePrimitives.push(this._currentPolygonShape);
        this.addPoint(x, y);
    }

    public lineTo(x: number, y: number) {
        this.addPoint(x, y);
    }

    public updateLinePoint(pointIndex: number, x: number, y: number) {
        let polygon: Polygon = this.shapePrimitives[0] as Polygon;
        polygon.points[pointIndex].x = x;
        polygon.points[pointIndex].y = y;
    }

    public insertAfterLinePoint(insertAfterpointIndex: number, x: number, y: number) {
        // let polygon:Polygon = shapePrimitives[0] as Polygon;
        // if (polygon -> points.capacity() < (insertAfterpointIndex < 0 ? 1 : insertAfterpointIndex + 1)) {
        //     polygon -> points.reserve(polygon -> points.capacity() + 10);
        // }
        // auto it = polygon -> points.begin() + insertAfterpointIndex + 1;
        // polygon -> points.insert(it, Vec2(x, y));
    }

    public deleteLinePoint(pointIndex: number) {
        let polygon: Polygon = this.shapePrimitives[0] as Polygon;
        // polygon.points.erase(polygon -> points.begin() + pointIndex);
    }

    private endShape(closed: boolean = false) {
        if (this._currentPolygonShape) {
            this._currentPolygonShape.closed = closed;
            this._currentPolygonShape = null;
        }
    }

    public close(): void {
        this.endShape(true);
    }

    public bezierCurveTo(ctX1: number, ctY1: number, ctX2: number, ctY2: number, endX: number, endY: number) {
        let last = this._currentPolygonShape.points[this._currentPolygonShape.points.length - 1];
        // casteljauBerzier2D(last.x, last.y, ctX1, ctY1, ctX2, ctY2, endX, endY, 0, _currentPolygonShape -> points);
    }

    public drawEllipse(cx: number, cy: number, rx: number, ry: number) {
        this.endShape();
        this.dirty = true;
        // this.shapePrimitives.push(new Ellipse(cx, cy, rx, ry));
    }

    public drawCircle(cx: number, cy: number, r: number) {
        this.endShape();
        this.dirty = true;
        // this.shapePrimitives.push(new Circle(cx, cy, r));
    }

    public drawRect(x: number, y: number, w: number, h: number) {
        this.endShape();
        this.dirty = true;
        // this.shapePrimitives.push(new Rectangle(x, y, w, h));
    }

    public drawCapsule(c1x: number, c1y: number, c2x: number, c2y: number, radius: number) {
        this.endShape();
        this.dirty = true;
        // this.shapePrimitives.push(new Capsule(c1x, c1y, c2x, c2y, radius));
    }

    //     public build(std:: vector<Vec2> & points, std:: vector < unsigned short > & triangles, float strokeWidth)
    // {
    //     for (int i = 0; i < shapePrimitives.size(); ++i)
    //     {
    //         if (action == GraphicsPathAction::FILL)
    //         {
    //             shapePrimitives[i] -> fill(points, triangles);
    //         }
    //         else
    //         {
    //             shapePrimitives[i] -> stroke(points, triangles, strokeWidth);
    //         }
    //     }
    // }

    // public fill()
    // {
    //     action = GraphicsPathAction:: FILL;
    // }

    // public stroke()
    // {
    //     action = GraphicsPathAction:: STROKE;
    // }

    public clear() {
        this.shapePrimitives.length = 0;
    }

    override sampleStrokePoints(dist: number, normal?: boolean, normalType?: number): Float32Array {
        let data = new Float32Array(2000);
        let currentNum: number = 0;
        for (let i = 0; i < this.shapePrimitives.length; ++i) {
            let points: Float32Array = this.shapePrimitives[i].sampleStrokePoints(dist, normal, normalType);
            let remainSize: number = data.length - currentNum;
            if (remainSize > points.length) {
                data.set(points, currentNum);
            } else {
                let data1 = new Float32Array(data.length * 2);
                data1.set(data);
                data1.set(points, currentNum);
            }
            currentNum += points.length;
        }
        if (currentNum < data.length) {
            data = data.slice(0, currentNum);
        }
        return data;
    }

    override buildFillForTexture(): Texture2d {
        throw new Error("Method not implemented.");
    }
}

File: native/engine/core/geom/2d/shapes/Path.cpp
#include "Path.h"
#include "../../curve/bezier2D.h"
#include "./Ellipse.h"
#include "./Rectangle.h"
#include "./Capsule.h"
#include "./Circle.h"

using namespace d2;

ShapePrimitive *Path::getShapeFromCache(ShapeType type)
{
    for (int i = 0; i < cacheShape.getCount(); ++i)
    {
        if (cacheShape[i]->shapeType == type)
        {
            ShapePrimitive *shape = cacheShape[i];
            cacheShape.removeSwap(i);
            return shape;
        }
    }
    return nullptr;
}
void Path::addPoint(float &x, float &y)
{
    _currentPolygonShape->points.push(Vec2(x, y));
}

void Path::moveTo(float x, float y)
{
    dirty = true;
    d2::Polygon *shape = (d2::Polygon *)getShapeFromCache(ShapeType::POLYGON);
    if (!shape)
    {
        shape = new d2::Polygon();
    }
    _currentPolygonShape = shape;
    shapePrimitives.push(_currentPolygonShape);
    addPoint(x, y);
}

void Path::lineTo(float x, float y)
{
    addPoint(x, y);
}

void Path::updateLinePoint(int pointIndex, float x, float y)
{
    d2::Polygon *polygon = dynamic_cast<d2::Polygon *>(shapePrimitives[0]);
    polygon->points[pointIndex].x = x;
    polygon->points[pointIndex].y = y;

    // printf("updateLinePoint %d %f %f \n",pointIndex,x,y);
}

void Path::insertAfterLinePoint(int insertAfterpointIndex, float x, float y)
{
    d2::Polygon *polygon = dynamic_cast<d2::Polygon *>(shapePrimitives[0]);
    polygon->points.insert(insertAfterpointIndex, Vec2(x, y));
}

void Path::deleteLinePoint(int pointIndex)
{
    d2::Polygon *polygon = dynamic_cast<d2::Polygon *>(shapePrimitives[0]);
    polygon->points.removeAt(pointIndex);
}

void Path::endShape(bool closed)
{
    if (_currentPolygonShape)
    {
        _currentPolygonShape->closed = closed;
        _currentPolygonShape = NULL;
    }
}

void Path::close()
{
    endShape(true);
}

void Path::bezierCurveTo(float ctX1, float ctY1, float ctX2, float ctY2, float endX, float endY)
{
    Vec2 &last = *_currentPolygonShape->points.getLast();
    casteljauBerzier2D(last.x, last.y, ctX1, ctY1, ctX2, ctY2, endX, endY, 0, _currentPolygonShape->points);
}

void Path::drawEllipse(float cx, float cy, float rx, float ry)
{
    endShape();
    dirty = true;

    d2::Ellipse *shape = (d2::Ellipse *)getShapeFromCache(ShapeType::ELLIPSE);
    if (!shape)
    {
        shape = new d2::Ellipse();
    }
    shape->set(cx, cy, rx, ry);
    shapePrimitives.push(shape);
}

void Path::drawCircle(float cx, float cy, float r)
{
    endShape();
    dirty = true;
    d2::Circle *shape = (d2::Circle *)getShapeFromCache(ShapeType::CIRCLEC);
    if (!shape)
    {
        shape = new d2::Circle();
    }
    shape->set(cx, cy, r);
    shapePrimitives.push(shape);
}

void Path::drawRect(float x, float y, float w, float h)
{
    endShape();
    dirty = true;

    d2::Rectangle *shape = (d2::Rectangle *)getShapeFromCache(ShapeType::RETANGLE);
    if (!shape)
    {
        shape = new d2::Rectangle();
    }
    shape->set(x, y, w, h);
    shapePrimitives.push(shape);
}

void Path::drawCapsule(float c1x, float c1y, float c2x, float c2y, float radius)
{
    endShape();
    dirty = true;
    d2::Capsule *shape = (d2::Capsule *)getShapeFromCache(ShapeType::CAPUSLE);
    if (!shape)
    {
        shape = new d2::Capsule();
    }
    shape->set(c1x, c1y, c2x, c2y, radius);
    shapePrimitives.push(shape);
}

void Path::build(sqstd::Array<Vec2> &points, sqstd::Array<unsigned short> &triangles, float strokeWidth)
{
    for (int i = 0; i < shapePrimitives.getCount(); ++i)
    {
        if (action == GraphicsPathAction::FILL)
        {
            (*shapePrimitives.get(i))->fill(points, triangles);
        }
        else
        {
            (*shapePrimitives.get(i))->stroke(points, triangles, strokeWidth);
        }
    }
}

void Path::fill()
{
    action = GraphicsPathAction::FILL;
}

void Path::stroke()
{
    action = GraphicsPathAction::STROKE;
}

void Path::fill(FillStyle &fillStyle)
{
    this->fillStyle = fillStyle;
    fill();
}

void Path::stroke(StrokeStyle &strokeStyle)
{
    this->strokeStyle = strokeStyle;
    stroke();
}

void Path::clear()
{
    for (int i = 0; i < shapePrimitives.getCount(); ++i)
    {
        ShapePrimitive *shape = *shapePrimitives.get(i);
        shape->reset();
        cacheShape.push(shape);
    }
    shapePrimitives.clear();
    action = GraphicsPathAction::NONE;
    fillStyle.reset();
    strokeStyle.reset();
}

void Path::getBound(Bound2 &out)
{
    out.set(0, 0, 0, 0);
    Bound2 b;
    for (int i = 0; i < shapePrimitives.getCount(); ++i)
    {
        shapePrimitives[i]->getBounds(b);
        out.addBound(b);
    }
}

bool Path::containsPoint(float x, float y)
{
    for (int i = 0; i < shapePrimitives.getCount(); ++i)
    {
        if (action == GraphicsPathAction::FILL)
        {
            if (shapePrimitives[i]->contains(x, y))
            {
                return true;
            }
        }
        else
        {
            if (shapePrimitives[i]->strokeContains(x, y, 2))
            {
                return true;
            }
        }
    }
    return false;
}

void Path::queryGemoetry(float x, float y, d2::QueryShapeResult &result)
{
    for (int i = 0; i < shapePrimitives.getCount(); ++i)
    {
        if (action == GraphicsPathAction::FILL)
        {
            shapePrimitives[i]->queryFill(x, y, result);
            if (result.gemoetryIndex != -1)
            {
                result.gemoetryIndex = i;
                return;
            }
        }
        else
        {
            shapePrimitives[i]->queryStroke(x, y, result, 2);
            if (result.linePoint1Index != -1)
            {
                result.gemoetryIndex = i;
                return;
            }
        }
    }
}

Path::~Path()
{
    for (int i = 0; i < shapePrimitives.getCount(); ++i)
    {
        delete *shapePrimitives.get(i);
    }
    shapePrimitives.clear();

    for (int i = 0; i < cacheShape.getCount(); ++i)
    {
        delete *cacheShape.get(i);
    }
    cacheShape.clear();
}